main/src/com/google/refine/InterProjectModel.java

```diff
@@ -158,11 +175,21 @@ protected void computeJoin(ProjectJoin join) {
         if (fromColumn == null || toColumn == null) {
             return;
         }
-        
+
         for (Row fromRow : fromProject.rows) {
-            Object value = fromRow.getCellValue(fromColumn.getCellIndex());
-            if (ExpressionUtils.isNonBlankData(value) && !join.valueToRowIndices.containsKey(value)) {
-                join.valueToRowIndices.put(value, new ArrayList<Integer>());
+            Object fromRowKey = fromRow.getCellValue(fromColumn.getCellIndex());
+            if (ExpressionUtils.isNonBlankData(fromRowKey)) {
+                Object[] fromRowKeys;
+                if (separatorRegexp != null && !separatorRegexp.isEmpty() && fromRowKey instanceof String) {
+                    fromRowKeys = ((String) fromRowKey).split(separatorRegexp);
```

Comment:
I think it would be more efficient to compile this regexp to a `Pattern` only once, outside the loop, and re-use that inside the loop.
https://stackoverflow.com/questions/1720191/java-util-regex-importance-of-pattern-compile#1721778